Fixed #30427, Fixed #16176 -- Corrected setting descriptor in Field.contribute_to_class().

jarekwg opened the pull request:
[Ticket #30427](https://code.djangoproject.com/ticket/30427)

Previously: DeferredAttributes would not get stapled onto models where the model (or an ancestor) already had an existing *truthy* attribute (despite comment only mentioning protecting against overriding *classmethods*).
With this change: DeferredAttributes will ~not get stapled onto models where the model (or an ancestor) already has a *callable* attribute~ always get stapled on, regardless of other attributes on the models or their ancestors.

This means we have the following changes:
- If a model field is declared twice (once on the class and once on the ancestor), mro order is respected when deciding which should take precedence.
- If a model field is declared over a class attribute on the ancestor, the model field will override.

Aside from the change, this PR includes:
- Failing tests. ~A failing test that this change fixes.~
- ~An update to an existing test, covering the new overriding order of DeferredAttributes. Arguably, what was previously being tested for didn't make sense.~

Problems:
- Albeit obscure, this introduces a slight change in behaviour, so we might have to somehow communicate that..
- ~`@property` decorators throw a fly in the ointment. They look like methods, but the decorator makes `callable` return `False` for them. It's why the testsuite is failing presently. We can add an additional check for the existence of `__isabstractmethod__` on the attribute, but it starts to feel rabbitholey. Perhaps we should think about a more aggressive change here, such as always overriding everything. Why do we protect classmethods anyway?~

Happy to discuss ideas on how to solve this here or on the ticket.

[2020-03-04] edit: updated to reflect most recent decisions/discussion.

carltongibson commented on django/db/models/fields/__init__.py:
```diff
@@ -737,7 +737,8 @@ def contribute_to_class(self, cls, name, private_only=False):
             # such fields can't be deferred (we don't have a check for this).
-            if not getattr(cls, self.attname, None):
+            existing_attr = getattr(cls, self.attname, None)
+            if existing_attr is None or not callable(existing_attr):
```
I guess I'd pull this into a helper, since you're going to have to add `@property` decorators (at least) as well.

jarekwg replied:
Hey Carlton,
Thanks for following up on this.
Will update soon.

mxschmitt commented:
@jarekwg @carltongibson

Any updates related to this issue? As far as I see has Jarek updated his PR.

https://github.com/jazzband/django-model-utils/issues/382 is e.g. waiting for a fix.

Thanks!

carltongibson commented on tests/model_inheritance/test_abstract_inheritance.py:
```diff
@@ -60,7 +61,7 @@ class Derived(DescendantOne, DescendantTwo):
         self.assertEqual(DescendantOne._meta.get_field('name').max_length, 30)
         self.assertEqual(DescendantTwo._meta.get_field('name').max_length, 50)
-        self.assertEqual(Derived._meta.get_field('name').max_length, 50)
+        self.assertEqual(Derived._meta.get_field('name').max_length, 30)
```
This change isn't right. `Derived.mro()` is 

```
[Derived, 
 DescendantOne, 
 DescendantTwo, 
 AbstractBaseOne, 
 AbstractBaseTwo, 
 django.db.models.base.Model, 
 object,
]
```

The existing test is correct that `Derived` should inherit the `name` field from `DescendantTwo`. i.e. `max_length` should be 50.

carltongibson replied:
OK, so re-reading the description: 

> An update to an existing test, covering the new overriding order of DeferredAttributes. Arguably, what was previously being tested for didn't make sense.

So, running this change we have: 

```
DescendantOne.__dict__['name']
<django.db.models.query_utils.DeferredAttribute object at 0x102869278>
```

This explains the change in the test, **but** it's not what we expect from Python inheritance. So... we're trading one issue for another. 🤔

jarekwg replied:
OOOOH. Gross, you're right.
I would've assumed order is:
```
[Derived, 
 DescendantOne,
 AbstractBaseOne,
 AbstractBaseTwo,
 DescendantTwo,
 django.db.models.base.Model, 
 object,
]
```
This would be correct if `DescendantTwo` wasn't also inheriting from `AbstractBaseOne` and `AbstractBaseTwo`.

Thanks for looking at this again! Let me know what you think after you've thought about it some more, but I think you're right that the change in this line should be reverted.. I'll need to check why it's messed up the order then.. hrm.

carltongibson replied:
Some time later... 🙂

About to mail the list re whether this change is acceptable.

jarekwg commented:
Awesome. Thanks for the detailed review!
- I agree that _always_ setting the descriptor is arguably a better way to go. I believe additional tests may fail as a result, and obviously the change may affect more users too, but I'm really not a fan of the `is_attname_settable` method I've currently got in this PR when we don't have a solid case for why we desire such specific behaviour.
- I will do a bit of my own investigation into with why mro is affected by this; the metaclasses probably have a hand in it. If the rabbithole gets too deep, it's nice you're saying we might be happy to concede that we won't support pure python MRO. The "diamond" case is kind of unusual anyway so usage of that would be pretty low.
- I'll read through the other tickets you've linked and see if I this change helps them; if yes, will add tests.
- Yes, I'm definitely interested in seeing this to completion, and your guidance will certainly help. Give me a few weeks to come back with the next iteration, and we'll discuss more then.

Cheers,

jarekwg commented:
- ~This ticket (you have one).~ -- yep
- ~jazzband/django-model-utils#382~ -- covered by my initial `test_shadow_parent_attribute_with_field` test:
```
> from django.contrib.auth.models import AbstractUser

# Prior to change in this PR:
> AbstractUser.is_active
True  # True is defined on AbstractBaseUser and was preventing AbstractUser from overriding it.

# After the change in this PR:
> AbstractUser.is_active
<django.db.models.query_utils.DeferredAttribute object at 0x7fa1d01a1b20>
```
- ~ticket-16176~ -- covered in `test_shadow_parent_property_with_field`
- ~ticket-28198~ -- arguably already covered by `test_shadow_parent_attribute_with_field` ensuring we've got an instance of DeferredAttribute, but I've added `ShadowDeferTests` to defer/test.py to cover more explicitly.
- ticket-27807 -- not relevant I think. `username_validator` does not come through the `contribute_to_class` method. I'd assume it inherits just fine, but the point at which the field is created happens prior to any inheritance behaviour. I've added a test for this as well (for your reference), though I'd be inclined to back it out and have it addressed separate to this work.

I've stuck to the wording 'shadow', since that's what was in there already, though i _think_ shadowing refers more to cross-scope variable overrides, and might not entirely be correct in this case.. lemme know if you think it should be changed.

As for the diamond case, I can see what's going on.
This is the line that adds fields for abstract base models: https://github.com/jarekwg/django/blame/ticket_30427/django/db/models/base.py#L274
We iterate over them in mro order, however when we ask them what fields they have to offer, we don't ask them if the fields were defined explicitly on them, or on one of their ancestors.
So in this case, `DescendantOne` comes in as the first abstract base, and even though it doesn't define `name` directly, it inherits it from `AbstractBaseTwo`, basically pulling it up and causing it to get looked at before the `name` defined directly on `DescendantTwo`. Once `name` is looked at once, all future abstract bases that try to bring it in are ignored.
The fix would be to have abstract bases only `add_to_class` fields that they define explicitly (not inherit from a parent), however this is actually nontrivial to check for (I believe). If we want to fix this, we'll have to get a little crafty, and I'm not sure it's worth the effort. Lemme know what you think!
Alternatively, we don't iterate over the base classes at all and only staple on fields at the top level (where python would've already ensured correct mro behaviour), however that's a much more drastic change/simplification, and there must be reasons it's been done the way it is now.

carltongibson commented:
Hi @jarekwg. Good stuff. I had a quick look and rebased and squashed. 

>  ...I'd be inclined to back it out ...

OK, do that. I agree. I was just being optimistic I think 🙂

I need to go through your comments and the failures one by one to determine if they're acceptable to change.

jarekwg commented:
Removed `username_validator` tests. Left a diff for them on https://code.djangoproject.com/ticket/27807#comment:15

carltongibson commented:
OK. Great stuff @jarekwg. I’ll look at this again next week now. 

But we’ll definitely need a release note for the breaking change. 

Then we’ll need to approach the list and ask for extra feedback. I think the big we’re fixing (the User case being the clearest in my mind) is probably worth it but it’s not something I’d take in without consultation first. (So, release notes, my review, mailing list...)

carltongibson commented on tests/invalid_models_tests/test_models.py:
```diff
@@ -914,10 +914,8 @@ def test_property_and_related_field_accessor_clash(self):
-                pass
-
+        # Override related field accessor.
+        Model.fk_id = property(lambda self: 'ERROR')
```
Yeah... 🙂

jarekwg commented on django/db/models/fields/__init__.py:
```diff
@@ -761,11 +761,7 @@ def contribute_to_class(self, cls, name, private_only=False):
-            # such fields can't be deferred (we don't have a check for this).
-            if not getattr(cls, self.attname, None):
-                setattr(cls, self.attname, self.descriptor_class(self))
+            setattr(cls, self.attname, self.descriptor_class(self))
```
We could:
- delay merging this work, and plop a warning/deprecation message in here instead, preparing people for the new behaviour. Arguably the urgency of the fix outweighs the niche change in field inheritance interaction.
- raise a warning (or bubble up into a django check) that the user has set an attribute with the same name twice and whatever they have on their parent is getting ignored.

However I think both cases would likely suffer from warning users even when the overriding they're doing is intentional -- the `AbstractUser` case I provided above is a prime example of one that would result in warnings that the user wouldn't be able to do anything about, because the parent class is defined in Django itself..

jarekwg commented:
Let me know if you'd rather have the commit squashed into previous. I'm leaving it separate to make it more obvious where I've made changes.

carltongibson commented:
Hey @jarekwg. Thanks for picking this up. Please do leave separate commits for now. (Exactly makes it easier to review.)

carltongibson commented:
Right, ["truthy" is a word](https://developer.mozilla.org/en-US/docs/Glossary/Truthy), so I added it to the spelling wordlist. 

I added an example to the release notes. It's a bit long but I don't see that it can be much shorter and still get the point across? (@felixxm @adamchainz?) 

If we're happy, I **think** we're there... 🤞

Thanks for your work on this one @jarekwg. It's a monster. 👹

adamchainz commented:
> It's a bit long but I don't see that it can be much shorter and still get the point across? (@felixxm @adamchainz?)

I can't see anything to cut out. I'd even consider adding a sentence that `makemigrations` will detect changes if they exist (they will right?).

felixxm commented on docs/releases/3.1.txt:
```diff
@@ -452,6 +452,50 @@ Miscellaneous
+            abstract = True
+
+    class Derived(ParentOne, ParentTwo):
+        pass
```
This example raises:
```
test_one.Derived.name: (models.E006) The field 'name' clashes with the field 'name' from model 'test_one.derived'.
```

carltongibson replied:
Good catch! 

I ran this exact test **but** defining the model classes dynamically in the test case. I want to check adding similar to a `models.py` to make sure/see what happens when the checks are run. 

 I need to think about this one a bit. Ref's https://github.com/django/django/pull/11337#discussion_r391534297

carltongibson commented:
Ah, didn't actually run the new test... Let me work out what it **did change** here. 

(Sigh 🙂)

carltongibson commented:
Right, this is the pull request that never dies. 

Putting the [old test](https://github.com/django/django/blob/193c109327c5216cab1d4eed6bfdff24629b09a3/tests/model_inheritance/test_abstract_inheritance.py#L37-L63) back in multiply inheriting fields **is enabled** by this change. 

The test added in 4bbe8261c402694a1da3efcaafe3332f9c57af15 fails, and the old test needs adjusting, as we had before, since normal Python MRO is not respected. 

@jarekwg Do have capacity to play with this and work out **exactly** what the ordering is etc? 

* This is [the relevant bit of `ModelBase.__new__()`](https://github.com/django/django/blob/master/django/db/models/base.py#L273-L278) It's looking at `new_class._meta.local_fields` at each point there and making sure it's sane. (I'll try and fire up the debugger for that again, but it keeps eating the time I allocate for it...) 
* Adjusting the failing test here. 
* (Re-)Adding a diamond-like example, like the old test, and adjusting that as we had it, with the comment
* Then documenting exactly what we can say about the behaviour. 

On `master` you can't multiply inherit a field (you've never been able to) so we're not breaking anything (AFAICS). We should be able to get a fix in (the new test cases are compelling). If not we should have a definite idea of what the blocker is.

jarekwg commented:
@carltongibson sure!
Just to clarify, what further insights (beyond what I summarised [here](https://github.com/django/django/pull/11337#issuecomment-588770574)) are you after when you ask:
>... work out exactly what the ordering is etc?

Resummarising my previous comment in short: abstract base classes are traversed in mro() order, but the way we then add fields from them onto the new class is reckless -- we don't check whether the fields are defined _directly_ on the classes we're traversing, or on one of their ancestors.

I'll:
- address the new failing test from master.
- re-add the diamond test.
- have another read of the `ModelBase.__new__()` code to see if we can get diamond inheritance working without making any scary changes to this part of the code.
- update docs accordingly, depending on the changes.

Will try to have a crack this long easter weekend.

carltongibson commented:
Hi Jared. 

I want to be 100% clear on the state of the model at each point. 

The old test looked like it was working. Except whilst it had a single Deferred descriptor, there were two identically named fields added (to local_fields) which is why, when we added the model check() call the error was raised. 

It wasn’t really the case that the fields were overridden in the sense of normal inheritance. Rather when you fetch the field you’d get the one from the base class, not the DescendantTwo. 

With the change, I want to make sure that the resultant model is sane. Maybe an added assert or two on the state of the _meta object makes sense. 

(As I say, I want to fire it up in the debugger myself as well.)

This is such a central part of the framework that we must tread very carefully. 

Unless we have no option, I don’t think further changes to `ModelBase.__new__()` is a good idea here.

jarekwg commented:
With my latest read-through of `ModelBase.__new__()`, I just noticed that it doesn't even do a full mro traversal of ancestors as i'd initially thought; it _only_ looks at direct ancestors due to the `base not in parents` condition here:
https://github.com/django/django/blob/6485a5f450b3dc60e690c31a75e0e9574a896842/django/db/models/base.py#L208

So there really is _absolutely no way_ of supporting diamond inheritance in the current state, because even if we did tell `DescendantOne` to only provide fields that were _directly_ defined on it, we'd never end up looping through to its ancestors (ie `AbstractBase`) to find the other fields..

It desperately needs a rewrite; dancing around it is just going to continue causing pain.
But for that to happen safely, we need beefier tests.

Soooo, I went a little more aggressive on these abstract inheritance tests, adding some utils for quick class creation, then applying a more rigorous covering of test cases. The PRO of this approach is we can write a tonne of complex inheritance cases in a readable way. The CON is a pattern like this could itself potentially introduce issues, or be a little harder to maintain. Please let me know if I've gone too far and I can revert the changes back to an expanded form. Please also let me know what you think of this suite of test cases; whether they're all sensible, or somewhy superfluous.

Also, what do you think about playing things out this way:
- We hold off with this PR (again), and introduce a new check, specifically for any instance of diamond inheritance (regardless of what kind of field definitions exist). If identified, slap user on wrist and tell them Django doesn't support this. If we give them a firm NO now, then we'll have grounds to fix it in future without worrying about causing further regressions.
- We merge this PR. Failing test for diamond inheritance will be a sleeper.
- One day, `ModelBase.__new__()` will get rewritten, at which point we can remove the check and the `@expectedfailure` off the test.

I made a commit changing tests only, cause would like to know what you think; but happy to update docs too.


re the sanity you were referring to, I had a quick look with:
```
                # Add fields from abstract base class if it wasn't overridden.
                print(base._meta)
                print(parent_fields)
                for field in parent_fields:
                    print(new_class.__dict__)
                    if (field.name not in field_names and
                            field.name not in new_class.__dict__ and
                            field.name not in inherited_attributes):
                        new_field = copy.deepcopy(field)
                        print(f'BEFORE: {new_class._meta.local_fields}')
                        new_class.add_to_class(field.name, new_field)
                        print(f'AFTER: {new_class._meta.local_fields}')
```
`new_class._meta.local_fields` appears to have the new field cleanly added, once only, thanks to the `field.name not in new_class.__dict__` check. Here's some sample output from running the above against one of my new tests:
```
--------------------------------------------------
DerivedFromBases12NO
model_inheritance.abstractbase1
[<django.db.models.fields.CharField: foo>]
{'__module__': 'model_inheritance.test_abstract_inheritance', '__doc__': None, '_meta': <Options for DerivedFromBases12NO>, 'DoesNotExist': <class 'model_inheritance.test_abstract_inheritance.DerivedFromBases12NO.DoesNotExist'>, 'MultipleObjectsReturned': <class 'model_inheritance.test_abstract_inheritance.DerivedFromBases12NO.MultipleObjectsReturned'>}
BEFORE: []
AFTER: [<django.db.models.fields.CharField: foo>]
model_inheritance.abstractbase2
[<django.db.models.fields.CharField: foo>]
{'__module__': 'model_inheritance.test_abstract_inheritance', '__doc__': None, '_meta': <Options for DerivedFromBases12NO>, 'DoesNotExist': <class 'model_inheritance.test_abstract_inheritance.DerivedFromBases12NO.DoesNotExist'>, 'MultipleObjectsReturned': <class 'model_inheritance.test_abstract_inheritance.DerivedFromBases12NO.MultipleObjectsReturned'>, 'foo': <django.db.models.query_utils.DeferredAttribute object at 0x7f85f00f8e50>}
--------------------------------------------------
```

adamchainz commented on tests/model_inheritance/test_abstract_inheritance.py:
```diff
@@ -1,63 +1,106 @@
+        self.assertCorrectInheritance('DerivedFromDescendantNONO', 30)
+        self.assertCorrectInheritance('DerivedFromDescendantNOO', 50)
+        self.assertCorrectInheritance('DerivedFromDescendantONO', 40)
+        self.assertCorrectInheritance('DerivedFromDescendantOO', 50)
```
Appreciate what you've done with the tests but I think they're harder to comprehend now. As a Django dev I can jump in and read model classes, but the helpers are obscuring things a bit to me.

Also you now have multiple tests per test method now - ideally there should just be a handful of classes per test and some assertions. I guess you can split based on overriding, non-overriding, etc.

jarekwg replied:
Thanks for taking a look!

Yeah it is more obscure, I was just hoping to find a way to avoid the verbosity of
```

Meta:
    abstract = True
```
over and over again.

Happy to revert, I guess the exercise helped list out several additional cases that weren't being covered before, like making sure each level pulls through from the previous in the absence of overrides.
I agree that test methods should hold minimal assertions that tie more directly to one specific use case, but I found it hard to break these down into multiple methods without having even more repetition/verbosity. Splitting on overriding/non-overriding still fans out a fair amount (if we care for each branch of the fan).

jarekwg replied:
Also, question: Is there a pattern for any kind of "short-form" class creation in the django test suite currently? I had a quick look but couldn't find anything, probs due to the "readability" argument.
Could be a worthwhile exercise to implement a standard for "quickbuild" setup functions for model classes -- would be more understandable if such things were used across the board, and much easier to scan than tonnes of vertical scrolling.

adamchainz replied:
I'm pretty sure you can make the abstract bases once in the test case class level:

```
class ...Tests(...):
    class SomeBase(models.Model):
         ...
```

Then reuse them appropriately in the individual test methods, which should only need to create the "bottom" classes?

I grepped for `\btype\(` and found a few uses, but none are for "quick" building.

I personally think the "tonnes of vertical scrolling" is a small concern compared to having to grok how `type()` works.

jarekwg replied:
> I'm pretty sure you can make the abstract bases once in the test case class level:
> 
> ```
> class ...Tests(...):
>     class SomeBase(models.Model):
>          ...
> ```
> 
> Then reuse them appropriately in the individual test methods, which should only need to create the "bottom" classes?

Ok, I like that. Wasn't aware you could declare there without having isolation or referencing issues, but had a quick play and looks pretty solid.

> I grepped for `\btype\(` and found a few uses, but none are for "quick" building.
> 
> I personally think the "tonnes of vertical scrolling" is a small concern compared to having to grok how `type()` works.

Well the tradeoff is you learn the extra bit of lingo (don't necessarily have to read up on `type`; docstring on util function gives arguably sufficient blackbox-level insight) to then be able to sweep through a whole bunch of class definitions quickly. Having them defined so close together allows for easy comparison between them, as well as between their usage across tests.
I think if someone were to visit this file in the future to work on a single test, then having everything splayed out as normal django class definitions is much more accessible. However, when looking at the tests holistically (to identify all the cases they're covering and how), a compact perspective can be useful.
I've seen a lot of django tests where multiple contributors have added test cases to the same suite, with approaches that are all quite different, with high potential for covering the same functionality multiple times. Seeing more neighbouring tests on the screen when writing one's own can lead to more consistency, both in test layout, and in ensuring the tests all work together to provide maximal non-overlapping coverage.
So I think there's at least _some_ merit to it. But I understand trying to forge a new pattern on a random PR probs serves only to detract from what the PR's discussion should actually be about.

Thanks for having a look anyway, I'll expand it out again after Carlton drops his thoughts on some of the other stuff I mentioned in my main comment (blocking diamond inheritance altogether).

adamchainz commented on tests/model_inheritance/test_abstract_inheritance.py:
```diff
@@ -1,63 +1,106 @@
+            NOTE: We need a rewrite of `ModelBase.__new__()` to support this properly.
+            It's marked as expected failure for now, and because we can't control
+            the inheritance the way we'd like to, we should block users from building
+            diamond cases at all, rather than testing for the converse.
```
> we should block users from building diamond cases at all, rather than testing for the converse.

But we don't block, right? It's not normally worth leaving comments about things Django "should" do.

Also I think this test can be split into a few different diamond cases.

jarekwg replied:
> > we should block users from building diamond cases at all, rather than testing for the converse.
> 
> But we don't block, right? It's not normally worth leaving comments about things Django "should" do.

Yeah fair. I'll update this comment once we decide whether we go ahead and address the `should` bit prior to merging this PR.

> Also I think this test can be split into a few different diamond cases.

There are 4 cases (A=Abstract, *=defines field). I think all are worth checking, based on :
```
  A*     A*     A*     A*
 / \    / \    / \    / \ 
A*  A  A*  A  A   A* A   A*
 \ /    \ /    \ /    \ /  
  D*     D      D*     D
```

So four separate test cases for diamond inheritance? Feels very verbose. Possibly no need to test in cases where `Derived` overrides with its own value since the earlier multi inheritance test already covers this, but given the complexity of the underlying metaclass logic, might be worth keeping still?

adamchainz replied:
I think it's fine. Separate test cases are normally a good thing since they can fail individually.

jarekwg commented:
> Sorry for the slow follow-up: in the first-place, 2020, but it's taken me a while to have a window to sit down with this again, and to get to stare at `ModelBase.__new__()` for sufficiently long. 😳

Heh, no worries. Thanks for circling back round to this.

I'll need to refresh my thoughts a bit again, but if you're saying that diamond-inheritance-wrist-slapping already occurs currently (not sure how i failed to notice this..), then we're good to go with pushing out this change. I'll try to find some time to expand out the tests this weekend.

Cheers,

carltongibson commented:
The key point is that such a model fails the `check()` call. (The old test didn't actually call `check()` so it was never caught...)

jarekwg commented:
> The key point is that such a model fails the `check()` call. (The old test didn't actually call `check()` so it was never caught...)

Ahh, ok that makes more sense. Thanks!

jarekwg commented on tests/model_inheritance/test_abstract_inheritance.py:
```diff
@@ -1,64 +1,184 @@
+    """ Test mro preserved when inheriting from multiple parents in diamond formation.
+        NO = non-overriding, O = overriding.
+
+        NOTE: We need a rewrite of `ModelBase.__new__()` to support this properly.
+        Until that happens, one of the tests here is marked as expected failure. Startup checks will
+        attempt to prevent users from building these cases anyway.
```
@carltongibson I've updated this note in response to @adamchainz' feedback, as well given what you said about the diamond check above. I think it's correct now and useful (?) to explain why we've marked one of the tests with `@expectedfailure`, but can remove it altogether if it still feels more like a ticket that's been shoved into the code.

aidanlister commented:
Great work on this @jarekwg ... @carltongibson how's it looking now?

carltongibson commented:
@aidanlister it's just waiting for another chance to review again. It's on my long list for 3.2.

carltongibson commented:
Closing in favour of #14508. Thanks @jarekwg!

jarekwg commented:
No worries! I'm just happy to see it get addresed.
